rclcpp/include/rclcpp/logger.hpp

```diff
@@ -150,20 +152,17 @@ class Logger
    * For example, ```get_logger('abc').get_child('def')``` will return a logger
    * with name `abc.def`.
    *
+   * It is recommended to use a temporary variable to store the return logger
+   * instead of calling it every time.
```

Comment:
I think it would be okay to call this every time, if sub-logger is already allocated, it just returns it from rcl. or do you want to mention this because of race condition to support multi-thread use case?

Reply:
I hope to be able to call it every time, but currently, it seems to be ineffective.
Consider the following situation,

```cpp
RCLCPP_INFO( node.get_child("child"), "a child message")
```

The `node.get_child("child")` will return a temporary variable `Logger` after creating a publisher using `rcl_logging_rosout_add_sublogger`. After publishing "a child message" by the created publisher, the temporary variable `Logger` will be immediately out of its scope, which means the publisher will be finalized. The problem is that it may happen that some external subscriptions might not receive this message.

I have two thoughts about fixing this issue:
1. not to destroy the publisher after Logger is destructed, make the publisher destroyed only if the corresponding node is destructed.
2. to destroy the publisher after Logger is destructed, but calling `rcl_publisher_wait_for_all_acked` for the `rosout` publisher before it's finalized.

It seems I was stuck here because I have no idea about which method could be accepted. If there are any suggestions, I'd appreciate that.

Reply:
> The problem is that it may happen that some external subscriptions might not receive this message.

I tested again and found that this problem not only occurs when the `rosout` publisher is destroyed immediately after publishing the message but also occurs when the message is sent before the external subscription information is collected on the DDS.

Reply:
@clalancette @fujitatomoya 

Creating a new publisher for each subordinate logger might not be the correct solution.

Reply:
Ah, right. you are right. it can be gone right away.

Reply:
How about destroy child when parent is destructed?

Reply:
> but also occurs when the message is sent before the external subscription information is collected on the DDS.

I think this subordinate logger can be just an item sharing the parent entry containing the existing publisher, we can't create a new publisher for this subordinate logger. I have updated https://github.com/ros2/rcl/pull/921, would you mind reviewing it again?

I'll remove this comment. `It is recommended to use a temporary variable to store the return logger ...`

Reply:
> it can be gone right away

Yes. Besides this, the publisher can't be used right away after it is newly created.

> How about destroy child when parent is destructed?

It only solves the former case (occurs when the rosout publisher is destroyed immediately), but not for the other case (the message is sent before the external subscription information is collected on the DDS.)

Reply:
Sorry about these comments above, which should be discussed on https://github.com/ros2/rcl/pull/921.

Reply:
> I think this subordinate logger can be just an item sharing the parent entry containing the existing publisher, we can't create a new publisher for this subordinate logger.

this sounds good instead of having the publisher for each one of them.

> but not for the other case (the message is sent before the external subscription information is collected on the DDS.)

why, i though that using parent publisher solved this problem too, because it should've been create before the subordinate logger?

Reply:
@iuhilnehc-ynos friendly ping. just checking if i am missing something.

Reply:
@fujitatomoya 

> i though that using parent publisher solved this problem too

Yes, using the parent publisher can solve this problem. I commented at [rcl](https://github.com/ros2/rcl/pull/921#discussion_r674481487) and had updated the [source code](https://github.com/ros2/rcl/pull/921/commits/f7a7a06b094ff8eaeb34d191067103087277a485).

And I also removed this comment of `rclcpp`.